### PR TITLE
Keras estimator: set reader's epoch = 1 to avoid sample duplication and drop-out in a single epoch

### PR DESCRIPTION
**Pull request by chongxiaoc**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description
We set infinite epochs for petastorm reader in horovod, due to the known issue that data partition is not guarantee to be equivalent size. TF's dataset shuffle function will automatically refill the shuffle buffer after retrieving a single element.
Code in horovod:
```
            if shuffle:
                dataset = dataset.shuffle(shuffle_buffer_size)
```
Mixing infinite iterations from reader and TF's shuffle function, will introduce sample duplication and drop-out in a single epoch. For example, training dataset is [`0,1,2,3,4,5,6,7]`, and shuffle size is 8.
Shuffling buffer is initialized as `[0,1,2,3,4,5,6,7]`
It randomly pick 7, and refill with 0 again: `[0,1,2,3,4,5,6,0]`
Now it is highly possible 0 will be selected twice in next few iterations, even before other values are selected once.

This PR fixes above issue by setting reader's epoch to be 1, while looping infinitely inside dataloader.
Also we introduced cache() support from `tf.data.Dataset` since it speeds up throughput.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by chongxiaoc** [comment]
Rebased and change pytorch inmem caching usage from petastorm `v0.11.0rc6`

**Review by tgaddair — APPROVED** [review]
LGTM!

**Comment by chongxiaoc** [comment]
Also fixes #2909

File: setup.py [hunk 1]
```diff
@@ -112,7 +112,7 @@ def build_extensions(self):
 pyspark_require_list = ['pyspark>=2.3.2;python_version<"3.8"',
                         'pyspark>=3.0.0;python_version>="3.8"']
 # Pin h5py: https://github.com/h5py/h5py/issues/1732
-spark_require_list = ['h5py<3', 'numpy', 'petastorm>=0.9.8,<0.11.0', 'pyarrow>=0.15.0']
+spark_require_list = ['h5py<3', 'numpy', 'petastorm>=0.11.0', 'pyarrow>=0.15.0']
```
Comment (EnricoMi) [inline setup.py:115]: @tgaddair do you think requiring latest petastorm for Horovod on Spark is too much of a constraint? Should we support petastorm <0.11 for a transition period?

If we go forward with this, that "limitation" must be recorded in the CHANGELOG and release notes.

Comment (chongxiaoc) [reply setup.py:115]: The only difference of `petastorm v0.11.0` is refactoring `pytorch inmem dataloader`, which I believe is not used frequently. Yep, I agree to add `CHANGELOG` and release notes.

Comment (chongxiaoc) [reply setup.py:115]: updated CHANGLOG to bring `pytorch inmem dataloder` change.

Comment (tgaddair) [reply setup.py:115]: I think this is fine. Petastorm is pretty lightweight with few overlapping dependencies, so shouldn't be an issue. Unfortunately, it's frequently the case that Horovod needs to be updated to the latest Petastorm to take advantage of new features.
